include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+ * currently pending buffers are still in use by the GPU), we have no choice but
+ * the free the buffer using the current stream.
+ *
+ * In either case, `max_reserve` represents the maxmum size of the stock. If
+ * returning a buffer would lead to the stock exeeding this quantity, the buffer
+ * is queued for destruction.
+ */
+template <class Allocator>
+class MemoryPool final {
+ public:
+  using pool_type = MemoryPool<Allocator>;
+  using alloc_type = typename Allocator::type;
+  template <class Container>
+  class Workspace {
+   public:
+    inline Workspace() : pool_{nullptr}, ctx_{nullptr} {}
+
+    inline Workspace(pool_type* pool, cudaStream_t ctx)
+        : pool_{pool}, ctx_{ctx} {}
+
+    Workspace(const Workspace&) = delete;
+    Workspace& operator=(const Workspace&) = delete;
+
+    inline Workspace(Workspace&& other)
+        : pool_{other.pool_},
+          ctx_{other.ctx_},
+          buffers_{std::move(other.buffers_)} {}
+
+    inline Workspace& operator=(Workspace&& other) {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), ctx_);
+      }
+      pool_ = other.pool_;
+      ctx_ = other.ctx_;
+      buffers_ = std::move(other.buffers_);
+      other.pool_ = nullptr;
+      return *this;
+    }
+
+    inline ~Workspace() {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), ctx_);
+      }
+    }
+
+    template <class T>
+    constexpr void at(const size_t n, T* ptr) const {
+      *ptr = at<T>(n);
+    }
+
+    template <class T>
+    constexpr T at(const size_t n) const {
+      return reinterpret_cast<T>(buffers_.at(n));
+    }
+
+    template <class T>
+    constexpr void get(const size_t n, T* ptr) const {
+      *ptr = get<T>(n);
+    }
+
+    template <class T>
+    constexpr T get(const size_t n) const {
+      return reinterpret_cast<T>(buffers_[n]);
+    }
+
+    constexpr alloc_type* operator[](const size_t n) const {
+      return buffers_[n];
+    }
+
+   protected:
+    pool_type* pool_;
+    cudaStream_t ctx_;
+    Container buffers_;
+  };
+
+  template <size_t N>
+  class StaticWorkspace final : public Workspace<std::array<alloc_type*, N>> {
+   public:
+    using base_type = Workspace<std::array<alloc_type*, N>>;
+
+    inline StaticWorkspace() : base_type() {}
+
+    inline StaticWorkspace(pool_type* pool, cudaStream_t ctx)
+        : base_type(pool, ctx) {
+      auto& buffers = this->buffers_;
+      pool->get_raw(buffers.begin(), buffers.end(), ctx);
+    }
+
+    StaticWorkspace(const StaticWorkspace&) = delete;
+    StaticWorkspace& operator=(const StaticWorkspace&) = delete;
+
+    inline StaticWorkspace(StaticWorkspace&& other)
+        : base_type(std::move(other)) {}
+    inline StaticWorkspace& operator=(StaticWorkspace&& other) {
+      base_type::operator=(std::move(other));
+      return *this;
+    }
+  };
+
+  class DynamicWorkspace final : public Workspace<std::vector<alloc_type*>> {
```

Comment:
The  `DynamicWorkspace` looks never be used, if yes, removing it would be better.

Reply:
It is used in tests. It exists to ensure that the API is generic.

Reply:
Not all code needs to fulfill a function right now.

Reply:
Make sense to me.